pages/en/indexing/operating-graph-node.mdx

```diff
@@ -0,0 +1,397 @@
+
+> Note that multiple Graph Nodes can all be configured to use the same database, which itself can be horizontally scaled via sharding.
+
+> TODO: recommendations on a basic setup, and then when / how to scale
+
+### Deployment rules
+
+Given multiple Graph Nodes, it is necessary to manage deployment of new subgraphs so that the same subgraph isn't being indexed by two different nodes, which would lead to collisions. This can be done by using deployment rules, which can also specify which `shard` a subgraph's data should be stored in, if database sharding is being used. Deployment rules can match on the subgraph name and the network that the deployment is indexing in order to make a decision.
+
+Example deployment rule configuration:
+```
+[deployment]
+[[deployment.rule]]
+match = { name = "(vip|important)/.*" }
+shard = "vip"
+indexers = [ "index_node_vip_0", "index_node_vip_1" ]
+[[deployment.rule]]
+match = { network = "kovan" }
+# No shard, so we use the default shard called 'primary'
+indexers = [ "index_node_kovan_0" ]
+[[deployment.rule]]
+match = { network = [ "xdai", "poa-core" ] }
+indexers = [ "index_node_other_0" ]
+[[deployment.rule]]
+# There's no 'match', so any subgraph matches
+shards = [ "sharda", "shardb" ]
+indexers = [
+    "index_node_community_0",
+    "index_node_community_1",
+    "index_node_community_2",
+    "index_node_community_3",
+    "index_node_community_4",
+    "index_node_community_5"
+  ]
+```
+
+Read more about deployment rules [here](https://github.com/graphprotocol/graph-node/blob/master/docs/config.md#controlling-deployment).
+
+### Dedicated query nodes
+
+Nodes can be configured to explicitly be query nodes by including the following in the configuration file:
+
+```
+[general]
+query = "<regular expression>"
+```
+
+Any node whose --node-id matches the regular expression will be set up to only respond to queries.
+
+### Database scaling via sharding
+
+For most use cases, a single Postgres database is sufficient to support a graph-node instance. When a graph-node instance outgrows a single Postgres database, it is possible to split the storage of graph-node's data across multiple Postgres databases. All databases together form the store of the graph-node instance. Each individual database is called a shard.
+
+Shards can be used to split subgraph deployments across multiple databases, and can also be used to use replicas to spread query load across databases. This includes configuring the number of available database connections each `graph-node` should keep in its connection pool for each database, which becomes increasingly important as more subgraphs are being indexed.
+
+> TODO: more guidance on when sharding might be useful, any pitfalls, recommendations
+
+> TODO: guidance on database connections?
```

Comment:
@lutter @leoyvens are there any references or pointers on database connection configuration?

Reply:
Sharding becomes useful when your existing database can't keep up with the load that `graph-node` puts on it, and when it's not possible to increase the database size anymore. I would always advise to first make a single database as big as possible before starting with shards. Although sharding can also help in situations where query traffic is split very unevenly between subgraphs; in those situations it can help dramatically if the high-volume subgraphs are kept in one shard and everything else in another because that setup makes it more likely that the data for the high-volume subgraphs stays in the db-internal cache and doesn't get replaced by data that's not needed as much from low-volume subgraphs.

As for db connections: I would start with `max_connections` in `postgresql.conf` set to 400 (or maybe even 200) and look at the `store_connection_wait_time_ms` and `store_connection_checkout_count` Prometheus metrics. Noticeable wait times (I'd say anything above 5ms) is an indication that there are too few connections available; high wait times there will also be caused by the database being very busy (like high CPU load) but if the database seems otherwise ok, high wait times indicate a need to increase the number of connections. In the configuration, how many connections each `graph-node` instance can use is an upper limit, and `graph-node` will not keep connections open if it doesn't need them.